dotnet/sdk-phase-1.md

```diff
@@ -0,0 +1,1052 @@
+* Phase 2 - Workflow worker
+* Phase 3 - Deterministic workflow sandbox and/or static analyzer
+
+This proposal is for phase 1.
+
+Notes for this proposal:
+
+* Intentionally loosely defined, everything subject to change
+* Some decisions are justified at the end
+* Lots of the code is not valid C# but has things like `/*...*/` for brevity or just empty implementations
+
+## Overview
+
+The .NET SDK will be backed by [sdk-core](https://github.com/temporalio/sdk-core), utilizing "P/Invoke" and a
+custom-built bridge.
+
+The high-level goals for the .Net SDK are:
+
+* Be as simple as possible from a caller POV
+* Be as lightweight as reasonable with regards to dependencies
+* Be reasonably flexible with reasonable defaults
+* Fit well into the .Net ecosystem
+* Minimum versions supported: .NET Core 3.1, .NET Standard 2.0, and .NET Framework 4.6.2
+* All common 64-bit platforms supported, 32-bit maybe later
+
+## Repository Management
+
+The existing repository at https://github.com/temporalio/sdk-dotnet has hundreds of commits, dozens of issues, open PRs,
+GH pages, etc that don't align with this proposal necessarily. After some thought, it has been decided to:
+
+* Create a new https://github.com/temporalio/sdk-dotnet-incubating repo for the very first commits
+* After that repo has some code and API doc gen capability:
+  * Rename https://github.com/temporalio/sdk-dotnet to https://github.com/temporalio/legacy-dotnet (note the
+    intentionally absent "sdk" term for SEO reasons)
+    * Add README note on https://github.com/temporalio/legacy-dotnet pointing to
+      https://github.com/temporalio/sdk-dotnet
+    * Disable GH pages
+    * Archive it
+    * Consider deletion at some point in the future
+  * Rename https://github.com/temporalio/sdk-dotnet-incubating to https://github.com/temporalio/sdk-dotnet
+  * Update DNS for https://dotnet.temporal.io/ to point to vercel-based API docs page
+
+Doing the repo rename was deemed easier than going through manually closing issues, closing PRs, creating empty `main`
+branch, and keeping `master` branch around for those reading older code.
+
+## Namespace/Package
+
+* Will publish as `Temporalio`
+  * Need to reserve prefix `Temporalio` on NuGet, see
+    https://learn.microsoft.com/en-us/nuget/nuget-org/id-prefix-reservation
+* Namespace will be `Temporal`
+  * See discussion topics later about this being `Temporal.Sdk` instead
+
+## Workflow Definition
+
+Here's a definition for an other-language-defined workflow (from Python README):
+
+```csharp
+namespace MyNamespace;
+
+using System.Text.Json.Serialization;
+using Temporal.Workflow;
+
+public record GreetingInfo
+{
+    [JsonPropertyName("salutation")]
+    public string Salutation { get; init; } = "Hello";
+
+    [JsonPropertyName("name")]
+    public string Name { get; init; } = "<unknown>";
+}
+
+[Workflow]
+public interface IGreetingWorkflow
+{
+    static readonly IGreetingWorkflow Ref = Refs<IGreetingWorkflow>.Instance;
```

Comment:
I didn't really tbh. The ref pattern is a result of research into referencing methods and the only way implemented after weighing it against the other two options (proxy lambda and source generation).

Reply:
> but the proxy lambda approach seems like a valid alternative.

Not sure I agree, but I'll explain in proposal. In Java we regret it because the call is literally on the proxy. In .NET, I guess I can ask people to please only use simple single calls in the lambdas and not do advanced things with it (we can't prevent all method calls on proxies).

If we forced people to make interfaces for everything they do, like Durable Entities and Temporal Java do, then it'd make more sense. I was hoping to avoid that. Proxies require virtual methods.

Worst case, we can add it on top.

Reply:
Added discussion point on lambda vs method reference